### PR TITLE
implemented string-valued text-overflow

### LINKED ISSUE #13709: Implement string-valued text-overflow in Servo
Part of #13702

In #13701, we added support for parsing the full range of values that `text-overflow` supports. However, [this is disabled in Servo mode](https://github.com/Manishearth/servo/blob/ea765688cbc7a7f5e223cf060f2bf1341dfa0ae5/components/style/properties/longhand/text.mako.rs#L15)

While supporting the two-value form of text-overflow (e.g. `text-overflow: "..." clip`) is cumbersome, it should be pretty straightforward to support the one-value form.

Steps:
- Remove the conditional in https://github.com/Manishearth/servo/blob/ea765688cbc7a7f5e223cf060f2bf1341dfa0ae5/components/style/properties/longhand/text.mako.rs#L15 , and use the `<%helpers:longhand ...` one unconditionally.
- Try compiling it. Layout will probably fail in [`push_fragment_to_line()` in `components/layout/inline.rs`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/inline.rs#L691), because `text_overflow::clip`/`text_overflow::ellipsis` are now `text_overflow::Side::Clip`/`text_overflow::Side::Ellipsis`, and what used to be an enum is now a struct. Match on `fragment.style().get_text().text_overflow.first` instead, and use  the new names for the variants.
- Make it possible to pass arbitrary strings to [`transform_into_ellipsis()`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/fragment.rs#L1013) as an argument. In `push_fragment_to_line()` add an extra match arm that is similar to the ellipsis one, but pushes a string. In the ellipsis one, pass "…".
- Test that the [lorem ipsum example on MDN works](https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow)

Relevant layout code: [`components/layout/inline.rs`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/inline.rs#L691) , [`components/layout/fragment.rs`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/fragment.rs#L1013)

Tests: None, but we will add some when this is done.

### PR DESCRIPTION
**Pull request by nikhilshagri**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13709 

- [x] There are tests for these changes

cc @Manishearth 
The lorem-ipsum example from MDN works as expected.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13924)

### CODE REVIEW

File: components/layout/inline.rs [hunk 1]
```diff
@@ -687,18 +687,23 @@ impl LineBreaker {
...
-        match (fragment.style().get_text().text_overflow,
-               fragment.style().get_box().overflow_x) {
-            (text_overflow::T::clip, _) | (_, overflow_x::T::visible) => {}
-            (text_overflow::T::ellipsis, _) => {
+
+        let text_overflow_string: Option<String> = match fragment.style().get_text().text_overflow.first {
+            longhands::text_overflow::Side::Clip => None,
+            longhands::text_overflow::Side::Ellipsis => {
                 need_ellipsis = fragment.margin_box_inline_size() > available_inline_size;
+                Some("...".to_string())
```
Comment (pcwalton) [inline components/layout/inline.rs]: This needs to be a Unicode ellipsis (copy and paste this: …)

**Comment by pcwalton** [comment]
Also, this needs a reftest. See `tests/wpt/mozilla/tests/css/text_overflow_a.html` for an example.

File: components/layout/fragment.rs [hunk 2]
```diff
@@ -1004,12 +1004,13 @@ impl Fragment {
...
     /// Transforms this fragment into an ellipsis fragment, preserving all the other data.
-    pub fn transform_into_ellipsis(&self, layout_context: &LayoutContext) -> Fragment {
+    pub fn transform_into_ellipsis(&self, layout_context: &LayoutContext, text_overflow_string: Option<String>)
+    -> Fragment {
         let mut unscanned_ellipsis_fragments = LinkedList::new();
         unscanned_ellipsis_fragments.push_back(self.transform(
                 self.border_box.size,
                 SpecificFragmentInfo::UnscannedText(
-                    box UnscannedTextFragmentInfo::new("…".to_owned(), None))));
+                    box UnscannedTextFragmentInfo::new(text_overflow_string.unwrap().to_owned(), None))));
```
Comment (emilio) [inline components/layout/fragment.rs]: no need for `to_owned` now here.

File: components/layout/inline.rs [hunk 3]
```diff
@@ -687,18 +687,23 @@ impl LineBreaker {
...
         let available_inline_size = self.pending_line.green_zone.inline -
             self.pending_line.bounds.size.inline - indentation;
-        match (fragment.style().get_text().text_overflow,
-               fragment.style().get_box().overflow_x) {
-            (text_overflow::T::clip, _) | (_, overflow_x::T::visible) => {}
-            (text_overflow::T::ellipsis, _) => {
+
+        let text_overflow_string: Option<String> = match fragment.style().get_text().text_overflow.first {
+            longhands::text_overflow::Side::Clip => None,
+            longhands::text_overflow::Side::Ellipsis => {
```
Comment (emilio) [inline components/layout/inline.rs]: can we write it like:

```rust
let need_ellipsis = match fragment.style().get_text().text_overflow.first {
    longhands::text_overflow::Side::Clip => false,
    longhands::text_overflow::Side::Ellipsis | longhands::text_overflow::Side::String(..) => ragment.margin_box_inline_size() > available_inline_size,
};
```

And then stop passing the string to `transform_into_ellipsis`, and look into `self.style()` there?

Also, this stops checking for `overflow-x: visible`, which I think is wrong, and would need a spec link if not.

**Review by emilio — CHANGES_REQUESTED** [review]

File: components/layout/inline.rs (hunk 3 above)
Comment (cynicaldevil) [inline components/layout/inline.rs]: @emilio I don't understand how looking into `self.style()` would help. Could you explain more?

Comment (emilio) [inline components/layout/inline.rs]: The same `match` you're doing now to get the string, you could do it inside `transform_into_ellipsis`.

That would keep the string manipulations in the place they belong to.

Comment (cynicaldevil) [inline components/layout/inline.rs]: Ah, got it. Thanks!

**Comment by cynicaldevil** [comment]
I've updated the PR with the changes;
@pcwalton I've modified the html test file to add a test for `string-text-overflow`, but don't we need a test for `clip` as well? It wasn't there originally, so I thought I would ask about it first.

File: components/layout/inline.rs [hunk 4]
```diff
@@ -684,16 +684,14 @@ impl LineBreaker {
...
-               fragment.style().get_box().overflow_x) {
-            (text_overflow::T::clip, _) | (_, overflow_x::T::visible) => {}
-            (text_overflow::T::ellipsis, _) => {
-                need_ellipsis = fragment.margin_box_inline_size() > available_inline_size;
-            }
-        }
+
+        let need_ellipsis = match fragment.style().get_text().text_overflow.first {
+            longhands::text_overflow::Side::Clip => false,
+            longhands::text_overflow::Side::Ellipsis |longhands::text_overflow::Side::String(..) =>
```
Comment (emilio) [inline components/layout/inline.rs]: please space after the pipe, also indent the following line.

File: tests/wpt/mozilla/tests/css/text_overflow_a.html [hunk 5]
```diff
@@ -6,15 +6,20 @@
...
+.overflow-string {
     overflow: hidden;
+    text-overflow: " [..]";
 }
 </style>
 </head>
 <body>
-<p id=hideme>mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
+<p class="overflow-ellipsis">mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
+<p class="overflow-string">mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/text_overflow_a.html]: With your changes, this tests nothing right now (it's a reftest that tests for "mismatch").

You should restore the original test and add a new one that tests for string.

File: components/layout/fragment.rs [hunk 6]
```diff
@@ -1004,12 +1004,18 @@ impl Fragment {
     }
 
     /// Transforms this fragment into an ellipsis fragment, preserving all the other data.
-    pub fn transform_into_ellipsis(&self, layout_context: &LayoutContext) -> Fragment {
+    pub fn transform_into_ellipsis(&self, layout_context: &LayoutContext)
+    -> Fragment {
         let mut unscanned_ellipsis_fragments = LinkedList::new();
+        let text_overflow_string: Option<String> = match self.style().get_text().text_overflow.first {
+            longhands::text_overflow::Side::Clip => None,
```
Comment (emilio) [inline components/layout/fragment.rs]: You should be able to make this path `unreachable!`, and return only the string in the two other paths, then remove the `unwrap()` after.

File: components/layout/inline.rs (hunk 4 above)
Comment (emilio) [inline components/layout/inline.rs]: Also, I believe you still want the check for `overflow_x: visible` (in which case`need_elipsis` should be false). Or at least leaving a comment about it, given it's unspecified.

**Review by emilio — CHANGES_REQUESTED** [review]

File: components/layout/inline.rs (hunk 4 above)
Comment (Manishearth) [inline components/layout/inline.rs]: Any reason why we're using need_ellipsis and an unwrap/unreachable instead of creating an Option here and restructuring it and passing it down to transform? IIRC this was the case in the PR originally, but it changed.

Comment (emilio) [inline components/layout/inline.rs]: @Manishearth I think it's cleaner to handle all the string-mangling inside `transform_into_ellipsis`, though feel free to disagree. In any case the overflow check/comment is still needed.

Comment (Manishearth) [inline components/layout/inline.rs]: I mean, we can avoid the double check on `overflow_x` by doing this.

```rust
let ellipsis = match ...first {
    Clip => None,
    _ if inline_size > available_size => None,
    Ellipsis => Some("...".to_owned()),
    String(s) => Some(s.to_owned()),
}

if let Some(ellipsis) = need_ellipsis {
    let ellipsis = fragment.transform_into_ellipsis(context, ellipsis /* this is a String */);
    ...
}
```
and in `transform_into_ellipsis` just apply the string directly.

Comment (Manishearth) [inline components/layout/inline.rs]: Oh, I meant `text_overflow`, sorry.

Comment (cynicaldevil) [inline components/layout/inline.rs]: I'm trying to do something like this:
```rust      
  let ellipsis = match (&fragment.style().get_text().text_overflow.first,
                                   fragment.style().get_box().overflow_x) {
            (&longhands::text_overflow::Side::Clip, _) | (_, overflow_x::T::visible) => None,

            (&longhands::text_overflow::Side::Ellipsis, _) |
            (&longhands::text_overflow::Side::String(ref string), _) => {
                match fragment.margin_box_inline_size() > available_inline_size {
                    true => {
                            // use the ref and return the appropriate Option<String>
                    },
                    false => None
                };
            }
        };
```
But the compiler complains that ref `string` is not bound in the `ellipsis` pattern. How do I get around this?

Comment (Manishearth) [inline components/layout/inline.rs]: Duplicate the match arm, use one for `Ellipsis` and one for `String`.

Comment (cynicaldevil) [inline components/layout/inline.rs]: That would result in a lot of duplicate code, would that be alright?

Comment (Manishearth) [inline components/layout/inline.rs]: Three lines of duplicate code. You don't even need the match:

``rust
if fragment.margin_box_inline_size() > available_inline_size { Some(string.to_owned()) } else { None }`
```
this much duplication is fine, and much more preferable to unreachable calls :smile:

**Comment by cynicaldevil** [comment]
@Manishearth @emilio I've updated the PR, and also changed the tests.

**Comment by emilio** [comment]
You still need to create another test, otherwise that test will always pass except *both* text overflows stop working.

**Comment by cynicaldevil** [comment]
So one for `string`, and another for `ellipsis`?

**Comment by emilio** [comment]
Yup!

**Comment by cynicaldevil** [comment]
@emilio Done!

File: components/layout/inline.rs [hunk 7]
```diff
@@ -684,21 +684,30 @@ impl LineBreaker {
...
         let available_inline_size = self.pending_line.green_zone.inline -
             self.pending_line.bounds.size.inline - indentation;
-        match (fragment.style().get_text().text_overflow,
-               fragment.style().get_box().overflow_x) {
-            (text_overflow::T::clip, _) | (_, overflow_x::T::visible) => {}
-            (text_overflow::T::ellipsis, _) => {
-                need_ellipsis = fragment.margin_box_inline_size() > available_inline_size;
+
+        let ellipsis = match (&fragment.style().get_text().text_overflow.first,
+                                   fragment.style().get_box().overflow_x) {
```
Comment (emilio) [inline components/layout/inline.rs]: nit: let's align these.

File: tests/wpt/mozilla/tests/css/text_overflow_ellipsis.html [hunk 8]
```diff
@@ -6,16 +6,17 @@
 p {
     width: 128px;
     background: gold;
-    text-overflow: ellipsis;
 }
-#hideme {
+p.hidden {
     overflow: hidden;
 }
+.overflow-ellipsis {
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/text_overflow_ellipsis.html]: Why? This doesn't apply `text-overflow: ellipsis` to both paragraphs as before.

File: tests/wpt/mozilla/tests/css/text_overflow_string.html [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+<link rel=mismatch href=text_overflow_ref.html>
+<style>
+p {
+    width: 128px;
+    background: gold;
+}
+p.hidden {
+    overflow: hidden;
+}
+.overflow-string {
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/text_overflow_string.html]: Same here, apply `text-overflow` to both paragraphs, then `overflow: hidden` to just one.

**Review by emilio — CHANGES_REQUESTED** [review]
You also need to add the tests and the rename to the manifest, have you run the tests?

**Comment by cynicaldevil** [comment]
Add the tests to manifest and also rectified other errors.

File: components/layout/fragment.rs [hunk 10]
```diff
@@ -1029,12 +1029,13 @@ impl Fragment {
     }
 
     /// Transforms this fragment into an ellipsis fragment, preserving all the other data.
-    pub fn transform_into_ellipsis(&self, layout_context: &LayoutContext) -> Fragment {
+    pub fn transform_into_ellipsis(&self, layout_context: &LayoutContext, text_overflow_string: String)
+    -> Fragment {
```
Comment (emilio) [inline components/layout/fragment.rs]: nit: Instead of this alignment, let's use

```rust
pub fn transform_into_ellipsis(&self,
			       layout_context: &LayoutContext,
			       text_overflow_string: String)
			       -> Fragment {
```

**Review by emilio — COMMENTED** [review]
Looks good to me with that, thanks!

**Comment by cynicaldevil** [comment]
I've changed the indentation, but tbh the older one looked better :P

**Comment by emilio** [comment]
@bors-servo r+

FWIW it's not about looking better (everyone has its opinion), but about being consistent inside the codebase.

Thanks for the patch!
